File: packages/server/src/scheduler.ts
import { readdir, rm } from "node:fs/promises";
import { spawnSync } from "node:child_process";
import { homedir } from "node:os";
import { join, resolve } from "node:path";

import { ApiError } from "./errors.js";
import { exists, readJsonFile } from "./utils.js";

export type ScheduledJobRun = {
  prompt?: string;
  command?: string;
  arguments?: string;
  files?: string[];
  agent?: string;
  model?: string;
  variant?: string;
  title?: string;
  share?: boolean;
  continue?: boolean;
  session?: string;
  runFormat?: string;
  attachUrl?: string;
  port?: number;
};

export type ScheduledJob = {
  scopeId?: string;
  timeoutSeconds?: number;
  invocation?: { command: string; args: string[] };

  slug: string;
  name: string;
  schedule: string;
  prompt?: string;
  attachUrl?: string;
  run?: ScheduledJobRun;
  source?: string;
  workdir?: string;
  createdAt: string;
  updatedAt?: string;
  lastRunAt?: string;
  lastRunExitCode?: number;
  lastRunError?: string;
  lastRunSource?: string;
  lastRunStatus?: string;
};

type JobEntry = {
  job: ScheduledJob;
  jobFile: string;
};

const SUPPORTED_PLATFORMS = new Set(["darwin", "linux"]);

function ensureSchedulerSupported() {
  if (SUPPORTED_PLATFORMS.has(process.platform)) return;
  throw new ApiError(400, "scheduler_unsupported", "Scheduler is supported only on macOS and Linux.");
}

function resolveHomeDir(): string {
  const home = homedir();
  if (!home) {
    throw new ApiError(500, "home_dir_missing", "Failed to resolve home directory");
  }
  return home;
}

function legacyJobsDir(): string {
  return join(resolveHomeDir(), ".config", "opencode", "jobs");
}

function schedulerScopesDir(): string {
  return join(resolveHomeDir(), ".config", "opencode", "scheduler", "scopes");
}

function legacyJobFilePath(slug: string): string {
  return join(legacyJobsDir(), `${slug}.json`);
}

function scopedJobFilePath(scopeId: string, slug: string): string {
  return join(schedulerScopesDir(), scopeId, "jobs", `${slug}.json`);
}

function normalizePathForCompare(value: string): string {
  const trimmed = value.trim();
  return trimmed ? resolve(trimmed) : "";
}

async function loadJobFile(path: string): Promise<ScheduledJob | null> {
  const job = await readJsonFile<Partial<ScheduledJob>>(path);
  if (!job || typeof job !== "object") return null;
  if (typeof job.slug !== "string" || typeof job.name !== "string" || typeof job.schedule !== "string") {
    return null;
  }
  return job as ScheduledJob;
}

async function loadLegacyJobEntries(): Promise<JobEntry[]> {
  const jobsDir = legacyJobsDir();
  if (!(await exists(jobsDir))) return [];
  const entries = await readdir(jobsDir, { withFileTypes: true });
  const jobs: JobEntry[] = [];
  for (const entry of entries) {
    if (!entry.isFile()) continue;
    if (!entry.name.endsWith(".json")) continue;
    const path = join(jobsDir, entry.name);
    const job = await loadJobFile(path);
    if (job) jobs.push({ job, jobFile: path });
  }
  return jobs;
}

async function loadScopedJobEntries(): Promise<JobEntry[]> {
  const scopes = schedulerScopesDir();
  if (!(await exists(scopes))) return [];
  const scopeEntries = await readdir(scopes, { withFileTypes: true });
  const jobs: JobEntry[] = [];
  for (const scopeEntry of scopeEntries) {
    if (!scopeEntry.isDirectory()) continue;
    const scopeId = scopeEntry.name;
    const jobsDir = join(scopes, scopeId, "jobs");
    if (!(await exists(jobsDir))) continue;
    const entries = await readdir(jobsDir, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isFile()) continue;
      if (!entry.name.endsWith(".json")) continue;
      const path = join(jobsDir, entry.name);
      const job = await loadJobFile(path);
      if (!job) continue;
      jobs.push({ job: { ...job, scopeId: job.scopeId ?? scopeId }, jobFile: path });
    }
  }
  return jobs;
}

async function loadAllJobEntries(): Promise<JobEntry[]> {
  const [scoped, legacy] = await Promise.all([loadScopedJobEntries(), loadLegacyJobEntries()]);
  return [...scoped, ...legacy];
}

function slugify(name: string): string {
  let out = "";
  let dash = false;
  for (const char of name.trim().toLowerCase()) {
    if (/[a-z0-9]/.test(char)) {
      out += char;
      dash = false;
      continue;
    }
    if (!dash) {
      out += "-";
      dash = true;
    }
  }
  return out.replace(/^-+|-+$/g, "");
}

function findJobEntryByName(entries: JobEntry[], name: string): JobEntry | null {
  const trimmed = name.trim();
  if (!trimmed) return null;
  const slug = slugify(trimmed);
  const lower = trimmed.toLowerCase();
  return (
    entries.find((entry) =>
      entry.job.slug === trimmed ||
      entry.job.slug === slug ||
      entry.job.slug.endsWith(`-${slug}`) ||
      entry.job.name.toLowerCase() === lower ||
      entry.job.name.toLowerCase().includes(lower)
    ) ?? null
  );
}

function schedulerSystemPaths(job: ScheduledJob): string[] {
  const home = resolveHomeDir();
  const paths: string[] = [];

  if (process.platform === "darwin") {
    if (job.scopeId) {
      paths.push(join(home, "Library", "LaunchAgents", `com.opencode.job.${job.scopeId}.${job.slug}.plist`));
    }
    paths.push(join(home, "Library", "LaunchAgents", `com.opencode.job.${job.slug}.plist`));
    return paths;
  }

  if (process.platform === "linux") {
    const base = join(home, ".config", "systemd", "user");
    if (job.scopeId) {
      paths.push(join(base, `opencode-job-${job.scopeId}-${job.slug}.service`));
      paths.push(join(base, `opencode-job-${job.scopeId}-${job.slug}.timer`));
    }
    paths.push(join(base, `opencode-job-${job.slug}.service`));
    paths.push(join(base, `opencode-job-${job.slug}.timer`));
    return paths;
  }

  return paths;
}

async function uninstallJob(job: ScheduledJob): Promise<void> {
  if (process.platform === "darwin") {
    for (const plist of schedulerSystemPaths(job)) {
      if (!(await exists(plist))) continue;
      spawnSync("launchctl", ["unload", plist]);
      await rm(plist, { force: true });
    }
    return;
  }

  if (process.platform === "linux") {
    const slug = job.slug;
    const scopedTimerUnit = job.scopeId ? `opencode-job-${job.scopeId}-${slug}.timer` : null;
    const legacyTimerUnit = `opencode-job-${slug}.timer`;
    for (const unit of [scopedTimerUnit, legacyTimerUnit].filter(Boolean) as string[]) {
      spawnSync("systemctl", ["--user", "stop", unit]);
      spawnSync("systemctl", ["--user", "disable", unit]);
    }

    for (const p of schedulerSystemPaths(job)) {
      if (await exists(p)) {
        await rm(p, { force: true });
      }
    }

    spawnSync("systemctl", ["--user", "daemon-reload"]);
    return;
  }

  ensureSchedulerSupported();
}

export async function listScheduledJobs(workdir?: string): Promise<ScheduledJob[]> {
  ensureSchedulerSupported();
  const entries = await loadAllJobEntries();

  const filterRoot = typeof workdir === "string" && workdir.trim() ? normalizePathForCompare(workdir) : null;
  const jobs = entries
    .map((entry) => entry.job)
    .filter((job) => {
      if (!filterRoot) return true;
      if (!job.workdir) return false;
      return normalizePathForCompare(job.workdir) === filterRoot;
    });

  jobs.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
  return jobs;
}

export async function resolveScheduledJob(
  name: string,
  workdir?: string
): Promise<{
  job: ScheduledJob;
  jobFile: string;
  systemPaths: string[];
}> {
  ensureSchedulerSupported();
  const trimmed = name.trim();
  if (!trimmed) {
    throw new ApiError(400, "job_name_required", "name is required");
  }

  const entries = await loadAllJobEntries();
  const filterRoot = typeof workdir === "string" && workdir.trim() ? normalizePathForCompare(workdir) : null;
  const filtered = filterRoot
    ? entries.filter((entry) => {
        const wd = entry.job.workdir;
        if (!wd) return false;
        return normalizePathForCompare(wd) === filterRoot;
      })
    : entries;

  const found = findJobEntryByName(filtered, trimmed);
  if (!found) {
    throw new ApiError(404, "job_not_found", `Job "${trimmed}" not found.`);
  }

  return {
    job: found.job,
    jobFile: found.jobFile,
    systemPaths: schedulerSystemPaths(found.job),
  };
}

export async function deleteScheduledJob(job: ScheduledJob, jobFile: string): Promise<void> {
  ensureSchedulerSupported();
  await uninstallJob(job);
  await rm(jobFile, { force: true });

  // Best-effort cleanup: if a legacy job file exists with the same slug, remove it.
  const legacy = legacyJobFilePath(job.slug);
  if (legacy !== jobFile && (await exists(legacy))) {
    await rm(legacy, { force: true });
  }
  // Best-effort cleanup: if a scoped job file exists with scopeId, remove it.
  if (job.scopeId) {
    const scoped = scopedJobFilePath(job.scopeId, job.slug);
    if (scoped !== jobFile && (await exists(scoped))) {
      await rm(scoped, { force: true });
    }
  }
}

File: packages/desktop/src-tauri/src/commands/scheduler.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::paths::home_dir;
use crate::types::ScheduledJob;

fn scheduler_supported() -> bool {
    cfg!(target_os = "macos") || cfg!(target_os = "linux")
}

fn require_scheduler_support() -> Result<(), String> {
    if scheduler_supported() {
        return Ok(());
    }
    Err("Scheduler is supported only on macOS and Linux.".to_string())
}

fn legacy_jobs_dir() -> Result<PathBuf, String> {
    let Some(home) = home_dir() else {
        return Err("Failed to resolve home directory".to_string());
    };
    Ok(home.join(".config").join("opencode").join("jobs"))
}

fn scheduler_scopes_dir() -> Result<PathBuf, String> {
    let Some(home) = home_dir() else {
        return Err("Failed to resolve home directory".to_string());
    };
    Ok(home
        .join(".config")
        .join("opencode")
        .join("scheduler")
        .join("scopes"))
}

fn normalize_path(input: &str) -> String {
    let trimmed = input.trim();
    if trimmed.is_empty() {
        return String::new();
    }
    if let Ok(abs) = fs::canonicalize(trimmed) {
        return abs.to_string_lossy().to_string();
    }
    trimmed.to_string()
}

fn load_job_file(path: &Path) -> Option<ScheduledJob> {
    let raw = fs::read_to_string(path).ok()?;
    serde_json::from_str(&raw).ok()
}

#[derive(Clone)]
struct JobEntry {
    job: ScheduledJob,
    job_file: PathBuf,
}

fn collect_legacy_jobs(jobs_dir: &Path) -> Vec<JobEntry> {
    let mut out = Vec::new();
    if !jobs_dir.exists() {
        return out;
    }
    let Ok(read_dir) = fs::read_dir(jobs_dir) else {
        return out;
    };
    for entry in read_dir.flatten() {
        let path = entry.path();
        if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        if let Some(job) = load_job_file(&path) {
            out.push(JobEntry {
                job,
                job_file: path,
            });
        }
    }
    out
}

fn collect_scoped_jobs(scopes_dir: &Path) -> Vec<JobEntry> {
    let mut out = Vec::new();
    if !scopes_dir.exists() {
        return out;
    }

    let Ok(scopes) = fs::read_dir(scopes_dir) else {
        return out;
    };

    for scope in scopes.flatten() {
        let scope_path = scope.path();
        if !scope_path.is_dir() {
            continue;
        }
        let scope_id = scope_path
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or("")
            .to_string();
        if scope_id.is_empty() {
            continue;
        }

        let jobs_dir = scope_path.join("jobs");
        if !jobs_dir.exists() {
            continue;
        }

        let Ok(entries) = fs::read_dir(&jobs_dir) else {
            continue;
        };

        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
                continue;
            }
            let Some(mut job) = load_job_file(&path) else {
                continue;
            };
            if job.scope_id.is_none() {
                job.scope_id = Some(scope_id.clone());
            }
            out.push(JobEntry {
                job,
                job_file: path,
            });
        }
    }

    out
}

fn slugify(name: &str) -> String {
    let mut out = String::new();
    let mut dash = false;
    for c in name.trim().to_lowercase().chars() {
        if c.is_ascii_alphanumeric() {
            out.push(c);
            dash = false;
            continue;
        }
        if !dash {
            out.push('-');
            dash = true;
        }
    }
    out.trim_matches('-').to_string()
}

fn find_job_entry_by_name(entries: &[JobEntry], name: &str) -> Option<JobEntry> {
    let trimmed = name.trim();
    if trimmed.is_empty() {
        return None;
    }
    let slug = slugify(trimmed);
    let lower = trimmed.to_lowercase();
    entries
        .iter()
        .find(|entry| {
            let job = &entry.job;
            job.slug == trimmed
                || job.slug == slug
                || job.slug.ends_with(&format!("-{slug}"))
                || job.name.to_lowercase() == lower
                || job.name.to_lowercase().contains(&lower)
        })
        .cloned()
}

fn collect_jobs_for_scope_root(scope_root: Option<&str>) -> Result<Vec<JobEntry>, String> {
    let legacy_dir = legacy_jobs_dir()?;
    let scopes_dir = scheduler_scopes_dir()?;

    let mut out = Vec::new();
    out.extend(collect_scoped_jobs(&scopes_dir));
    out.extend(collect_legacy_jobs(&legacy_dir));

    let filter_root = scope_root.map(|s| normalize_path(s)).unwrap_or_default();
    if !filter_root.is_empty() {
        out.retain(|entry| {
            entry
                .job
                .workdir
                .as_deref()
                .map(|wd| normalize_path(wd) == filter_root)
                .unwrap_or(false)
        });
    }

    out.sort_by(|a, b| a.job.name.to_lowercase().cmp(&b.job.name.to_lowercase()));
    Ok(out)
}

#[cfg(target_os = "macos")]
fn uninstall_job(slug: &str, scope_id: Option<&str>) -> Result<(), String> {
    let Some(home) = home_dir() else {
        return Err("Failed to resolve home directory".to_string());
    };

    let mut plists = Vec::new();
    if let Some(scope_id) = scope_id {
        let label = format!("com.opencode.job.{scope_id}.{slug}");
        plists.push(
            home.join("Library")
                .join("LaunchAgents")
                .join(format!("{label}.plist")),
        );
    }
    let legacy_label = format!("com.opencode.job.{slug}");
    plists.push(
        home.join("Library")
            .join("LaunchAgents")
            .join(format!("{legacy_label}.plist")),
    );

    for plist in plists {
        if !plist.exists() {
            continue;
        }
        let _ = Command::new("launchctl").arg("unload").arg(&plist).output();
        let _ = fs::remove_file(&plist);
    }

    Ok(())
}

#[cfg(target_os = "linux")]
fn uninstall_job(slug: &str, scope_id: Option<&str>) -> Result<(), String> {
    let Some(home) = home_dir() else {
        return Err("Failed to resolve home directory".to_string());
    };

    let base = home.join(".config").join("systemd").join("user");

    let mut timer_units: Vec<String> = Vec::new();
    if let Some(scope_id) = scope_id {
        timer_units.push(format!("opencode-job-{scope_id}-{slug}.timer"));
    }
    timer_units.push(format!("opencode-job-{slug}.timer"));

    for timer_unit in timer_units {
        let _ = Command::new("systemctl")
            .args(["--user", "stop", timer_unit.as_str()])
            .output();
        let _ = Command::new("systemctl")
            .args(["--user", "disable", timer_unit.as_str()])
            .output();
    }

    let mut files: Vec<PathBuf> = Vec::new();
    if let Some(scope_id) = scope_id {
        files.push(base.join(format!("opencode-job-{scope_id}-{slug}.service")));
        files.push(base.join(format!("opencode-job-{scope_id}-{slug}.timer")));
    }
    files.push(base.join(format!("opencode-job-{slug}.service")));
    files.push(base.join(format!("opencode-job-{slug}.timer")));

    for file in files {
        if file.exists() {
            let _ = fs::remove_file(&file);
        }
    }

    let _ = Command::new("systemctl")
        .args(["--user", "daemon-reload"])
        .output();
    Ok(())
}

#[cfg(not(any(target_os = "macos", target_os = "linux")))]
fn uninstall_job(_slug: &str, _scope_id: Option<&str>) -> Result<(), String> {
    Err("Scheduler is supported only on macOS and Linux.".to_string())
}

#[tauri::command]
pub fn scheduler_list_jobs(scope_root: Option<String>) -> Result<Vec<ScheduledJob>, String> {
    require_scheduler_support()?;
    let entries = collect_jobs_for_scope_root(scope_root.as_deref())?;
    Ok(entries.into_iter().map(|e| e.job).collect())
}

#[tauri::command]
pub fn scheduler_delete_job(
    name: String,
    scope_root: Option<String>,
) -> Result<ScheduledJob, String> {
    require_scheduler_support()?;
    let trimmed = name.trim();
    if trimmed.is_empty() {
        return Err("name is required".to_string());
    }

    let entries = collect_jobs_for_scope_root(scope_root.as_deref())?;
    let entry = find_job_entry_by_name(&entries, trimmed)
        .ok_or_else(|| format!("Job \"{trimmed}\" not found."))?;

    uninstall_job(&entry.job.slug, entry.job.scope_id.as_deref())?;
    if entry.job_file.exists() {
        fs::remove_file(&entry.job_file).map_err(|e| format!("Failed to remove job file: {e}"))?;
    }

    // Best-effort cleanup of any duplicates.
    if let Ok(legacy_dir) = legacy_jobs_dir() {
        let legacy = legacy_dir.join(format!("{}.json", entry.job.slug));
        if legacy != entry.job_file && legacy.exists() {
            let _ = fs::remove_file(legacy);
        }
    }
    if let (Some(scope_id), Ok(scopes_dir)) =
        (entry.job.scope_id.as_deref(), scheduler_scopes_dir())
    {
        let scoped = scopes_dir
            .join(scope_id)
            .join("jobs")
            .join(format!("{}.json", entry.job.slug));
        if scoped != entry.job_file && scoped.exists() {
            let _ = fs::remove_file(scoped);
        }
    }

    Ok(entry.job)
}
